src/query/interpreter.cpp

```diff
@@ -2061,6 +2126,54 @@ PreparedQuery PrepareIsolationLevelQuery(ParsedQuery parsed_query, const bool in
       RWType::NONE};
 }
 
+Callback SwitchMemoryDevice(storage::StorageMode current_mode, storage::StorageMode requested_mode,
+                            InterpreterContext *interpreter_context) {
+  Callback callback;
+  callback.fn = [current_mode, requested_mode, interpreter_context]() mutable {
+    if (current_mode == requested_mode) {
+      return std::vector<std::vector<TypedValue>>();
+    }
+    if (SwitchingFromDiskToInMemory(current_mode, requested_mode)) {
+      throw utils::BasicException(
+          "You cannot switch from on-disk storage to in-memory storage while the database is running. "
+          "Please restart the database to switch to in-memory storage.");
+    }
+    if (SwitchingFromInMemoryToDisk(current_mode, requested_mode)) {
+      std::unique_lock main_guard{interpreter_context->db->main_lock_};
+
+      if (auto vertex_cnt_approx = interpreter_context->db->GetInfo().vertex_count; vertex_cnt_approx > 0) {
+        main_guard.unlock();
+        throw utils::BasicException(
+            "You cannot switch from in-memory storage to on-disk storage when the database "
+            "contains data. Please delete all entries from the database, run FREE MEMORY and then repeat this "
+            "query. ");
+      }
+
+      main_guard.unlock();
+      if (interpreter_context->interpreters->size() > 1) {
+        throw utils::BasicException(
+            fmt::format("You cannot switch from in-memory storage to on-disk storage when there are "
+                        "multiple sessions active. Please close all other sessions and try again. {}",
+                        interpreter_context->interpreters->size()));
+      }
+
+      auto db_config = interpreter_context->db->config_;
+      interpreter_context->db = std::make_unique<memgraph::storage::DiskStorage>(db_config);
+    }
+    return std::vector<std::vector<TypedValue>>();
+  };
+  return callback;
```

Comment:
So not 100% sure on which exactly issues you think can happen but I understand what you meant when the session is created in the middle. I handled that in a different way. Each accessor gets created with some storage mode and then upon commit we check whether this is still the active mode, if not abort. I think this is better than having interpreter_context lock because otherwise you would need to lock Access method from the storage class, at least this was my conclusion but am willing to discuss that in more details.

Reply:
But here we are checking to see if we are the only active session, right?
We check if there is another interpreter and throw if we aren't the only one.
But there is a case were another session could be created between our check and the storage switch (I think).